src/monster.cpp

```diff
@@ -142,83 +153,88 @@ void monster::poly(mtype *t)
     }
     def_chance = type->def_chance;
     faction = t->default_faction;
+    upgrades = t->upgrades;
 }
 
-bool monster::can_upgrade() const
-{
-    // No upgrade_min, no upgrades ever
-    if (type->upgrade_min <= 0) {
-        return false;
-    }
-    // Hallucinations don't upgrade!
-    if (is_hallucination()) {
-        return false;
+bool monster::can_upgrade() {
+    return upgrades && (ACTIVE_WORLD_OPTIONS["MONSTER_UPGRADE_FACTOR"] > 0.0);
+}
+
+// For master special attack.
+void monster::hasten_upgrade() {
+    if (!can_upgrade() || upgrade_time < 1) {
+        return;
     }
-    // No chance of upgrading, abort
-    if ((type->half_life <= 0 && type->base_upgrade_chance <= 0) ||
-        (type->upgrade_group == mongroup_id( "GROUP_NULL" ) && type->upgrades_into == "NULL")) {
-        return false;
+
+    const int scaled_half_life = type->half_life * ACTIVE_WORLD_OPTIONS["MONSTER_UPGRADE_FACTOR"];
+    upgrade_time -= rng(1, scaled_half_life);
+    if (upgrade_time < 0) {
+        upgrade_time = 0;
     }
-    // Turned off means turned off
-    if (ACTIVE_WORLD_OPTIONS["MONSTER_UPGRADE_FACTOR"] <= 0) {
-        return false;
+}
+
+// This will disable upgrades in case max iters have been reached.
+// Checking for return value of -1 is necessary.
+int monster::next_upgrade_time() {
+    const int scaled_half_life = type->half_life * ACTIVE_WORLD_OPTIONS["MONSTER_UPGRADE_FACTOR"];
+    int day = 0;
+    for (int i = 0; i < UPGRADE_MAX_ITERS; i++) {
+        if (one_in(2)) {
```

Comment:
Occasional very small values are just fine, that's what we expect to
happen, you just need to set your decay rate accordingly.
When loading the new json values, they need a default applied, otherwise
people's preexisting monsters will never upgrade.  This wrecked my test run
since it was on a preexisting world.